### PR TITLE
[DPE-3007] Add Support to use PostgreSQL as Hive Metastore

### PR DESCRIPTION
**Pull request by theoctober19th**
Add support to use PostgredSQL as Hive Metastore by integrating `kyuubi-k8s` with `postgresql-k8s`.

### CODE REVIEW

**Review by welpaolo — COMMENTED** [review]
The PR looks good. I left some minor comments.
I believe that we are missing some integrations tests:
- Removal of integration with PostgreSQL and check that everything is working
- Integrating again with PostgreSQL and check the existence of the previously created tables and everything is working properly

File: src/charm.py [hunk 1]
```diff
@@ -153,6 +183,22 @@ def s3_connection_info(self) -> Optional[S3ConnectionInfo]:
             bucket=raw_info.get("bucket"),
         )
 
+    @property
+    def db_connection_info(self) -> Optional[DatabaseConnectionInfo]:
```
Comment (deusebio) [inline src/charm.py:187]: Following  @welpaolo comment, it is probably best to rename the property to refer to `metastore`

File: charmcraft.yaml [hunk 2]
```diff
@@ -56,6 +68,9 @@ resources:
...
     upstream-source: ghcr.io/canonical/charmed-spark:3.4-22.04_edge
 
+
 requires:
   s3-credentials:
-    interface: s3
\ No newline at end of file
+    interface: s3
+  database:
+    interface: postgresql_client
```
Comment (deusebio) [inline charmcraft.yaml:76]: **todo** we should probably put a limit on the relations. I don't think we could have more than 1 metastore

File: src/charm.py [hunk 3]
```diff
@@ -40,10 +50,13 @@ def __init__(self, *args):
         super().__init__(*args)
         self.workload = KyuubiServer(self.unit.get_container(KYUUBI_CONTAINER_NAME))
         self.s3_requirer = S3Requirer(self, S3_INTEGRATOR_REL)
+        self.database = DatabaseRequires(
```
Comment (deusebio) [inline src/charm.py:53]: maybe we could call it `self.metastore`

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -152,6 +157,103 @@ async def test_jdbc_endpoint(ops_test: OpsTest, test_pod):
...
+        cursor.execute(""" SELECT * FROM "DBS" WHERE "NAME" != 'default' """)
+        num_dbs = cursor.rowcount
+        cursor.execute(""" SELECT * FROM "TBLS" """)
+        num_tables = cursor.rowcount
+
+    connection.close()
+
+    # Assert that new database and tables have indeed been added to metastore
+    assert num_dbs != 0
+    assert num_tables != 0
```
Comment (deusebio) [inline tests/integration/test_charm.py:254]: **todo** Can we check also that when we remove the relation, we fall back on using the derby hive metastore?

File: src/charm.py [hunk 5]
```diff
@@ -53,12 +66,25 @@ def register_event_handlers(self):
...
         """Handle the `on_install` event."""
         self.unit.status = Status.WAITING_PEBBLE.value
 
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        logger.info("Database created...")
+        self.update_service()
+
+    def _on_database_relation_removed(self, event) -> None:
+        logger.info("Database relation removed")
+        self.update_service()
```
Comment (deusebio) [inline src/charm.py:86]: **todo** can we do a double check that this works as expected? both manually and with integration tests? The reason is that database removal events are somewhat curious, in that the relation still exists but it about to be torn down. What this means is that the data is still there and often in the logic we do checks on relation and/or data existence, thus not updating the service by removing the properties from the configuration files, until another `update_status` event is triggered.

This is actually the reason why in some removed events (in the new pattern, BTW at a certain point we should refactor the charm into the new pattern), we force the setting to "None" (see [here](https://github.com/canonical/spark-history-server-k8s-operator/blob/main/src/events/s3.py#L53)), to mimic that the relation was torn down (although the data is still there).

**Review by deusebio — COMMENTED** [review]
The changes look already rather good! So I believe we should be close. There is one point that I'd like to investigate to make sure that the behaviour is as expected. 

If it is has expected, it could be that this is abstracted by the DP libs (by [this](https://github.com/canonical/kyuubi-k8s-operator/blob/postgresql-integration/lib/charms/data_platform_libs/v0/data_interfaces.py#L802) function). If so, this function seems actually quite nice and worth to be used more when creating data-models in the `core` submodule.

File: charmcraft.yaml [hunk 6]
```diff
@@ -56,6 +68,9 @@ resources:
     description: OCI image for kyuubi
     upstream-source: ghcr.io/canonical/charmed-spark:3.4-22.04_edge
 
+
 requires:
   s3-credentials:
-    interface: s3
\ No newline at end of file
+    interface: s3
+  database:
```
[the remark this answers is not in the document]
Comment (theoctober19th) [reply charmcraft.yaml:75]: How about `metastore-db`, just like `auth-db` for authentication database?

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -6,10 +6,13 @@
...
+import psycopg2
 import pytest
 import yaml
 from pytest_operator.plugin import OpsTest
 
+from constants import METASTORE_DATABASE_NAME
+
 logger = logging.getLogger(__name__)
 
 APP_NAME = "kyuubi-k8s"
```
[the remark this answers is not in the document]
Comment (theoctober19th) [reply tests/integration/test_charm.py:18]: I think APP_NAME can be inferred from `charmcraft.yaml` file itself and `BUCKET_NAME` can be removed, thus eliminating the necessity for a constant file.

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -152,6 +157,103 @@ async def test_jdbc_endpoint(ops_test: OpsTest, test_pod):
...
+    kyuubi_unit = ops_test.model.applications[APP_NAME].units[0]
+    action = await kyuubi_unit.run_action(
+        action_name="get-jdbc-endpoint",
+    )
+    result = await action.wait()
+
+    jdbc_endpoint = result.results.get("endpoint")
+    logger.info(f"JDBC endpoint: {jdbc_endpoint}")
+
+    pod_name = test_pod
```
[the remark this answers is not in the document]
Comment (theoctober19th) [reply tests/integration/test_charm.py:198]: Updated.

File: src/charm.py (hunk 3 above)
Comment (theoctober19th) [reply src/charm.py]: Changed to `self.metastore_db`. The other for the authentication will be `self.auth_db`.

File: charmcraft.yaml (hunk 2 above)
Comment (theoctober19th) [reply charmcraft.yaml:76]: Updated to `limit: 1`.

File: src/charm.py (hunk 1 above)
Comment (theoctober19th) [reply src/charm.py]: Updated to `metastore_db_connection_info`.

File: src/charm.py (hunk 5 above)
Comment (theoctober19th) [reply src/charm.py:90]: During my tests, I've always found the `hive-site.xml` is updated by the relation broken event and never encountered the edge case you mentioned. I've also added a test case to check the removal of relation in the integration tests.

**Comment by theoctober19th** [comment]
> The PR looks good. I left some minor comments. I believe that we are missing some integrations tests:
> 
> * Removal of integration with PostgreSQL and check that everything is working
> * Integrating again with PostgreSQL and check the existence of the previously created tables and everything is working properly

I have added a test for 1. However, as we discussed yesterday, since we do not guarantee the persistence / removal of metadata upon the change of metastore from A to B to A back again, I have skipped 2 at the moment. WDYT @welpaolo @deusebio ?

**Comment by deusebio** [comment]
That's fine by me! We can maybe just have a tech debt in the backlog for this though. The use case that Paolo is suggesting is relevant, but a nice to have at the moment

**Comment by welpaolo** [comment]
Fine with me as well!

**Review by welpaolo — APPROVED** [review]

**Review by deusebio — APPROVED** [review]
All good!
